### ISSUE TITLE
How should business logic be tested when using decorators?

### ISSUE THREAD

**Issue by KaloyanP** [slackapi/bolt-python#245]
The approach of using decorators makes it much easier to get started. We have built an app using the examples as a reference implementation. However, we have hit issues with figuring out how to test the business logic wrapped by decorators.

E.g. for AWS Lambda, the example approach is roughly the following:

```python
from slack_bolt import App

app = App(token="xoxb-token", secret="secret")
@app.event({"type": "message", "subtype": None})
def process_message(message, say, request):
    # some more code goes here
    return do_some_logic_based_on_message

def handler(event, context):
    slack_handler = SlackRequestHandler(app=app)
    return slack_handler.handle(event, context)
```
Usually, the problem of isolating external dependencies for the purpose of unit testing can be resolved by splitting the business logic out into a separate class and injecting the Slack client into it, [then using mocks for testing like it's done in the tests](https://github.com/slackapi/bolt-python/blob/f2f8532c2d363ceb6ae97c1d60253ccb59dba933/tests/scenario_tests/test_events.py#L492). However, since the decorators get applied at module import, and in this case they refer to instance methods on the `app` object rather than functions in the `slack_bolt.app` module, the `app` object needs to be instantiated as a module variable before the `process_message` function definition.

This makes it difficult to set up a BusinessLogic class and pass a reference to `app` in the constructor because the methods on the class would be defined before `app` can be injected into the object instance.

Another option would be to create an instance of a BusinessLogic class after `app` has been created, then decorate static functions that call methods of that class. The reference to `app` would have to be passed into the object to allow it to post messages outside of the usual response - e.g. post to channels, etc. For example, like this:

```python
from slack_bolt import App

app = App(token="xoxb-token", secret="secret")
business_logic = BusinessLogic(app)
@app.event({"type": "message", "subtype": None})
def process_message(message, say, request):
    return business_logic.do_some_logic_based_on_message(message)

def handler(event, context):
    slack_handler = SlackRequestHandler(app=app)
    return slack_handler.handle(event, context)
```

Finally, there is always the option of just calling the bare app.* method without the decorators, which avoids some of the extra lines that get added by the decorators:

```python
from slack_bolt import App

app = App(token="xoxb-token", secret="secret")
business_logic = BusinessLogic(app)
# Register handler directly
app.event({"type": "message", "subtype": None}, business_logic.do_some_logic_based_on_message)

def handler(event, context):
    slack_handler = SlackRequestHandler(app=app)
    return slack_handler.handle(event, context)
```

I realise this is a bit rambley, and all of these options seem inelegant and clumsy, so I was wondering if there were some way of testing business logic that I am not seeing.

**Comment by seratch** [comment]
Hi @KaloyanP, thanks for writing in! 

As you've already checked the tests in this project, preparing fixture data for imitating requests from Slack and run `app.dispatch()` method with the data would be the safest way to verify the behavior of your apps. Also, we know that Bolt framework does not provide a great solution for unit tests yet. I cannot tell when we can deliver the feature but I'm planning to work on the addition in the near future.

For the module structure design for testing, indeed, there are several way to go. One possible easy way would be having `register_listeners(App)` method and using it in `main.py` (the entry point of your app). With this approach, you can separate the `App` initialization from listener methods and whatever you have for the application logic. 

Here is a pseudo code demonstrating a possible structure:

```python
# -----------------
# functions.py

def handle_hi_command(ack, body):
    ack("Hi there!")

# -----------------
# listeners.py

from slack_bolt import App
from .functions import handle_hi_command

def register_listeners(app: App):
    app.command("/hi")(handle_hi_command)

# -----------------
# main.py

from slack_bolt import App
from .listeners import register_listeners

app = App()

register_listeners(app)

if __name__ == "__main__":
    app.start(port=3000)


# -----------------
# test_functions.py

from .functions import handle_hi_command

def test_handle_hi_command(self):
    ack, body = MockAck(), {}  # prepare inputs
    result = handle_hi_commands(ack, body)
    assert something is True

# -----------------
# test_listeners.py

from slack_bolt import App
from .listeners import register_listeners

def test_app_listeners(self):
    app = App() # initialize App for testing
    register_listeners(app)
    response = app.dispactch(build_hi_command_request())
    assert something is True
```

Also, my personal app can be a hint for this topic. https://github.com/seratch/send-it-later-for-slack/tree/main/app

**Comment by seratch** [comment]
👋 Let us know if you need further help here. I will close this issue after waiting for your response for a few more days.

**Comment by KaloyanP** [comment]
Hi, just coming here to say that what @seratch posted worked perfectly - we ended up refactoring our app into that structure.

There is the small issue that we solved with an "awful hack" where the Slack mock API server is not included in the slack-bolt Python library, and different package managers for Python handle repo imports differently, so we couldn't use the mainline version that's maintained together with the bolt framework. Hopefully, if/once the bolt testing framework is ready at some point in the future, we can ditch this hack and it will work. 

The solution was to copy two of the test setup files, `utils` and `mock_web_server_api.py` to the test package of our project and modify them to work in a different folder structure. The imports in `mock_web_server_api.py` needed changing as the relative imports no longer worked.

Here are the two files we moved:
- mock_web_server_api.py - https://github.com/slackapi/bolt-python/blob/f2f8532c2d363ceb6ae97c1d60253ccb59dba933/tests/mock_web_api_server.py
- utils.py https://github.com/slackapi/bolt-python/blob/f2f8532c2d363ceb6ae97c1d60253ccb59dba933/tests/utils.py

And the folder structure we ended up with is:

```
app
├── app_package/
│   └── listeners.py #This is where the listeners get registered
│   └── functions.py #Business logic
└── test_app_package/
    └── slack_test_utils/
    │   ├── mock_web_server_api.py
    │   └── utils.py
    └── test.py #This is where the tests live
```

**Comment by latacora-tomekr** [comment]
@seratch how would you handle the mocking here in the case where `handle_hi_command` ends up using the SDK to get more data from the Slack API (for example, if the `handle_hi_command` returned the results of [conversations.history](https://api.slack.com/methods/conversations.history/code))? Is that a response you would have to handle and return in [mock_web_server_api.py](https://github.com/slackapi/bolt-python/blob/main/tests/mock_web_api_server.py#L19)?

**Comment by seratch** [comment]
@latacora-tomekr 
The main reason why this project's unit tests have the mock HTTP server is that we want to verify the library's behavior by performing real HTTP requests, not with only mock objects.

If you would like to have a mock object for `WebClient` in your tests, you can pass the mocked instance to the `App` constructor instead. With this approach, you can use any mock library for it. If this approach does not work for you, you can reuse the mock HTTP server in this project for now.

In the long run, we are planning to provide testing tools for Bolt apps (we cannot tell when we can release it, though). If you find obstacles and/or get insights through your Slack app development, sharing your knowledge (in a new issue) would be greatly appreciated.
